Code coverage automation script

adityapatwardhan opened the pull request:
This script downloads the latest nightly build package and executes the
tests on it. It also downloads all necessary tools and then uploads the
results to Coveralls.io. The badge for coverage is posted on README.md

adityapatwardhan committed:
Code coverage automation script

This script downloads the latest nightly build package and executes the
tests on it. It also downloads all necessary tools and then uploads the
results to Coveralls.io. The badge for coverage is posted on README.md

lzybkr commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,109 @@
+$openCoverZip = 'https://ci.appveyor.com/api/projects/PowerShell/powershell-f975h/artifacts/OpenCover.zip'
+
+
+New-Item "$PSScriptRoot\logs.txt"  -Force
```
You might consider putting the log in some more typical output directory, or at least adding the file you create to `.gitignore`.

lzybkr commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,109 @@
+
+try
+{
+    "Starting downloads." | Tee-Object -FilePath "$PSScriptRoot\logs.txt" -Append
```
Maybe you should add a `log` helper function.

lzybkr commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,109 @@
+
+    "Test run done.!!" | Tee-Object -FilePath "$PSScriptRoot\logs.txt" -Append
+
+    $powershellVersionString = (Get-Content "$psBinPath\publish\powershell.version")
```
I expect this file to go away at some point - it would be better to get the string from PowerShell itself, e.g.

```powershell
$powershellVersionString = & $psBinPath\publish\powershell.exe -noprofile -command '$PSVersionTable.GitCommitId'
```

adityapatwardhan committed:
Addressed code review comments and added readme.md file

Addressed CR comments about log file location, logging function and usage
of powershell.version file. Also added a README.md file explaining the
process.

adityapatwardhan commented:
@lzybkr  Addressed CR comments

adityapatwardhan committed:
Fixed bug in logging

adityapatwardhan committed:
Added upload completed message

adityapatwardhan committed:
Fixed get git version commandline bug

Fix the bug in invoking $PSversionTable. Added -force to expand-archive
cmdlets.

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+        $Path = "$env:Temp\CodeCoverageRunLogs.txt"
+    )
+
+    $message = "$([Datetime]::Now.ToShortDateString()) - $([datetime]::Now.ToShortTimeString()) : $message"
```
>$message = "$([Datetime]::Now.ToShortDateString()) - $([datetime]::Now.ToShortTimeString()) : $message" [](start = 4, length = 103)

this can also be done as:
`"{0:d} - {0:t} : {1}" -f ([datetime]::now),$message`
or even:
`"{0} : {1}" -f ([datetime]::now),$message`

adityapatwardhan replied:
Fixed.

---
In reply to: [94473993](https://github.com/PowerShell/PowerShell/pull/2918#discussion_r94473993) [](ancestors = 94473993)

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+    else
+    {
+        Add-Content -Path $Path -Value $Message -PassThru
+    }
```
would just `add-content -force` do the trick (add-content will create the file if it doesn't exist)

adityapatwardhan replied:
Fixed.

---
In reply to: [94474364](https://github.com/PowerShell/PowerShell/pull/2918#discussion_r94474364) [](ancestors = 94474364)

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+
+    Invoke-WebRequest -uri $codeCoverageZip -outfile "$outputBaseFolder\PSCodeCoverage.zip"
+    Invoke-WebRequest $testContentZip -outfile "$outputBaseFolder\tests.zip"
+    Invoke-WebRequest $openCoverZip -outfile "$outputBaseFolder\OpenCover.zip"
```
for consistency - add `-uri` to these lines too?

adityapatwardhan replied:
Fixed.

---
In reply to: [94474590](https://github.com/PowerShell/PowerShell/pull/2918#discussion_r94474590) [](ancestors = 94474590)

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+Write-LogPassThru -Message "openCoverZip: $openCoverZip"
+
+$outputBaseFolder = "$env:Temp\CC"
+$null = New-Item -ItemType Directory -Path $outputBaseFolder -Force
```
this path may already exist with contents. Would it be a good thing to delete the directory (recursively) if it exists before starting?

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+    $coverallsPath = "$outputBaseFolder\coveralls"
+
+    ## Saving the nupkg as zip so we can expand it.
+    Invoke-WebRequest $coverallsToolsUrl -outfile "$outputBaseFolder\coveralls.zip"
```
>$coverallsToolsUrl [](start = 22, length = 18)

`-uri`?

adityapatwardhan replied:
fixed

---
In reply to: [94474891](https://github.com/PowerShell/PowerShell/pull/2918#discussion_r94474891) [](ancestors = 94474891)

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+    Write-LogPassThru -Message $commitId
+
+    $coverallsPath = "$outputBaseFolder\coveralls"
+    $coverallsToken = 'SaRLabTUDNqaO2JT9uuYcR78a7kia044D'
```
secrets should probably be provided rather than embedded

adityapatwardhan replied:
Fixed.

---
In reply to: [94475097](https://github.com/PowerShell/PowerShell/pull/2918#discussion_r94475097) [](ancestors = 94475097)

JamesWTruher commented on test/tools/CodeCoverageAutomation/Start-CodeCoverageRun.ps1:
```diff
@@ -0,0 +1,129 @@
+
+    Write-LogPassThru -Message "Upload complete."
+}
+catch
```
I wonder if `ErrorActionPreference` should be set to `STOP`

adityapatwardhan replied:
Fixed.

---
In reply to: [94475385](https://github.com/PowerShell/PowerShell/pull/2918#discussion_r94475385) [](ancestors = 94475385)

adityapatwardhan committed:
Addressed CR feedback

JamesWTruher commented:
:shipit:

lzybkr committed:
Fix spelling

adityapatwardhan committed:
Merge branch 'master' into CodeCoverageAutomation

adityapatwardhan commented:
@lzybkr Merge conflict is solved. Change is ready to merge.

lzybkr committed:
Correct outdated step in README.md

adityapatwardhan commented:
@lzybkr All checks have passed. Is it ok to merge?
